cpp/src/Slice/SliceUtil.cpp

```diff
@@ -470,59 +470,49 @@ Slice::checkIdentifier(const string& id)
     {
         if(name.find(suffixBlacklist[i], name.size() - suffixBlacklist[i].size()) != string::npos)
         {
-            unit->error("illegal identifier `" + name + "': `" + suffixBlacklist[i] + "' suffix is reserved");
-            isValid = false;
+            unit->warning(ReservedIdentifiers, "`" + suffixBlacklist[i] +
+                                               "' suffix is reserved by Ice: the identifier `" + name +
+                                               "' may conflict with generated code");
         }
     }
 
+    // Check the identifier for illegal ice prefixes
+    if(unit->currentIncludeLevel() == 0 && ciequals(name.substr(0, 3), "ice"))
+    {
+        unit->warning(ReservedIdentifiers, "`" + name.substr(0, 3) + "' prefix is reserved by Ice: the identifier `" +
+                                           name + "' may conflict with generated code");
+    }
+
     // check the identifier for illegal underscores
     size_t index = name.find('_');
     if(index == 0)
     {
-        unit->error("illegal leading underscore in identifier `" + name + "'");
+        unit->error("illegal identifier: `" + name + "', leading underscores are reserved by Ice");
         isValid = false;
     }
     else if(name.rfind('_') == (name.size() - 1))
     {
-        unit->error("illegal trailing underscore in identifier `" + name + "'");
+        unit->error("illegal identifier: `" + name + "', trailing underscores are reserved by Ice");
         isValid = false;
     }
     else if(name.find("__") != string::npos)
     {
-        unit->error("illegal double underscore in identifier `" + name + "'");
+        unit->error("illegal identifier: `" + name + "', double underscores are reserved by Ice");
         isValid = false;
     }
-    else if(index != string::npos && unit->currentIncludeLevel() == 0 && !unit->allowUnderscore())
+    else if(index != string::npos && unit->currentIncludeLevel() == 0)
     {
-        DefinitionContextPtr dc = unit->currentDefinitionContext();
-        assert(dc);
-        if(dc->findMetaData("underscore") != "underscore") // no 'underscore' file metadata
-        {
-            unit->error("illegal underscore in identifier `" + name + "'");
-            isValid = false;
-        }
-    }
-
-    // Check the identifier for illegal ice prefixes
-    if(unit->currentIncludeLevel() == 0 && !unit->allowIcePrefix() && name.size() > 2)
-    {
-        DefinitionContextPtr dc = unit->currentDefinitionContext();
-        assert(dc);
-        if(dc->findMetaData("ice-prefix") != "ice-prefix") // no 'ice-prefix' metadata
-        {
-            if(ciequals(name.substr(0, 3), "ice"))
-            {
-                unit->error("illegal identifier `" + name + "': `" + name.substr(0, 3) + "' prefix is reserved");
-                isValid = false;
-            }
-        }
+        unit->warning(ReservedIdentifiers, "underscores are reserved by Ice: the identifier `" + name +
```

Comment:
I think the warning should be `identifiers with underscores are reserved ...`.

Reply:
Where would we actually say what the identifier was then?
Something like:
`identifiers with underscores are reserved by Ice: the identifier 'foo' may conflict with generated code`
is longer and sounds clunkier than
`underscores are reserved by Ice: the identifier 'foo' may conflict with generated code`

Reply:
The first sentence should use the term `identifier`, so for example:

_identifiers_ with underscores are _reserved_ by Ice: using `f_oo` as an identifier may result in name collisions in the generated code